Expose additional functions of the crypto engine to user

twisted-pear opened the pull request:
# What's new

- Allow use of the standard `furi_hal_crypto_encrypt` and `furi_hal_crypto_decrypt` functions with user supplied keys instead of using keys from the secure enclave. User supplied key can be loaded with `furi_hal_crypto_load_key` and unloaded with `furi_hal_crypto_unload_key`.
- Using user supplied keys, also allow use of the crypto engine's CTR and GCM mode. This can be done using `furi_hal_crypto_ctr` and `furi_hal_crypto_gcm`. For GCM there are convenience functions provided for encryption and decryption (`furi_hal_crypto_gcm_encrypt_and_tag` and `furi_hal_crypto_gcm_decrypt_and_verify`).
- Introduce a define called `FURI_HAL_CRYPTO_ADVANCED_AVAIL` which indicates the availability of these new functions.
- This is somewhat related to Issue #2885. It is not intended to affect the secure enclave in any way.

# Verification 

- Use of `furi_hal_crypto_encrypt` and `furi_hal_crypto_decrypt` remains unchanged. Data encrypted using a user supplied key and `furi_hal_crypto_encrypt` can be decrypted using the same key and `furi_hal_crypto_decrypt`.
- The output of the CTR function is intended to be identical to the output generated by the openssl CLI tool.
- The output of the GCM function is intended to be identical to the output generated by the GCM functions in this repo: https://github.com/mko-x/SharedAES-GCM
- I wrote a FAP automating all these tests which is available here: https://github.com/twisted-pear/crypto_test. This app will test CTR and GCM with a few preconfigured test vectors. It writes the results to the log at level Info.

# Checklist (For Reviewer)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

Commit:
Allow loading user supplied keys and add CTR mode

Commit:
Add GCM mode to furi_hal_crypto

Commit:
Split up CTR and GCM code, add flag for adv crypto

Commit:
Add convenience functions for GCM crypto

Commit:
Run fbt format

hedger committed:
Merge branch 'dev' into crypto_enhancements

Commit:
Update API symbols

twisted-pear commented:
Updated the API version and symbols this time so the automated builds can run through

twisted-pear commented:
How do I get around this:
```
API versions aren't matching for available targets. Please update!
==> firmware/targets/f18/api_symbols.csv <==
entry,status,name,type,params
Version,+,34.3,,

==> firmware/targets/f7/api_symbols.csv <==
entry,status,name,type,params
Version,+,34.4,,
```
Can I build for f18 myself somehow? Should I just update the api_symbols.csv manually?

Commit:
Update other API symbols

hedger commented:
@twisted-pear just run `./fbt TARGET_HW=18`.

twisted-pear commented:
thanks, i already figured that out
should build successfully now

Commit:
Revert "Update API symbols"

This reverts commit 4b06bb7af1aa30dc4ad5d392f1f38295dc2f9af5.

Commit:
Revert "Update other API symbols"

This reverts commit 4bad90748e74207118896d9f736a4157ec28c5f8.

Commit:
Update GCM to support additional auth data

Commit:
Update APIs

Commit:
Merge remote-tracking branch 'upstream/dev' into crypto_enhancements

skotopes committed:
Merge branch 'dev' into crypto_enhancements

skotopes committed:
FuriHal: update crypto documentation, method names and usage

skotopes commented on firmware/targets/f7/furi_hal/furi_hal_crypto.c:
```diff
@@ -276,6 +289,35 @@ bool furi_hal_crypto_store_unload_key(uint8_t slot) {
+    furi_assert(furi_hal_crypto_mutex);
+    furi_check(furi_mutex_acquire(furi_hal_crypto_mutex, FuriWaitForever) == FuriStatusOk);
+
+    furi_hal_bus_enable(FuriHalBusAES1);
```
I'd set AES_CR_EN and never ever touch it till key unloaded.

skotopes replied:
Except mode change

twisted-pear replied:
The original code also sets CR_EN once en/decryption starts. In GCM it also needs to be re-enabled when switching between certain phases.

skotopes replied:
k, make sense.

skotopes commented on firmware/targets/f7/furi_hal/furi_hal_crypto.c:
```diff
@@ -276,6 +289,35 @@ bool furi_hal_crypto_store_unload_key(uint8_t slot) {
+
+    furi_hal_bus_enable(FuriHalBusAES1);
+
+    if(!furi_hal_bt_is_alive()) {
```
Only if enclave used

twisted-pear replied:
done in new commit

skotopes commented on firmware/targets/f7/furi_hal/furi_hal_crypto.c:
```diff
@@ -340,3 +382,377 @@ bool furi_hal_crypto_decrypt(const uint8_t* input, uint8_t* output, size_t size)
+
+    furi_hal_bus_enable(FuriHalBusAES1);
+
+    if(!furi_hal_bt_is_alive()) {
```
same

skotopes commented on firmware/targets/f7/furi_hal/furi_hal_crypto.c:
```diff
@@ -340,3 +382,377 @@ bool furi_hal_crypto_decrypt(const uint8_t* input, uint8_t* output, size_t size)
+    size_t i;
+    for(i = 0; i < length - last_block_bytes; i += CRYPTO_BLK_LEN) {
+        if(!furi_hal_crypto_process_block_bswap(&input[i], &output[i], CRYPTO_BLK_LEN)) {
+            goto out_enc_err;
```
We prohibit use of `goto`, instead we use `do{ }while(0)`.Moment when you see issue with do-while usually means that complexity of your code is not acceptable anymore.

twisted-pear replied:
done in new commit

skotopes commented on firmware/targets/f7/furi_hal/furi_hal_crypto.c:
```diff
@@ -340,3 +382,377 @@ bool furi_hal_crypto_decrypt(const uint8_t* input, uint8_t* output, size_t size)
+    /* GCM header phase */
+
+    MODIFY_REG(AES1->CR, AES_CR_GCMPH, CRYPTO_GCM_PH_HEADER);
+    SET_BIT(AES1->CR, AES_CR_EN);
```
looks like you do want to make mode change method that disables CR_EN, changes mode, and then re-enables CR_EN

twisted-pear replied:
I do not. On some mode changes disabling CR_EN results in an invalid output according to the manual.

skotopes replied:
k, make sense then

skotopes commented on firmware/targets/f7/furi_hal/furi_hal_crypto.c:
```diff
@@ -340,3 +382,377 @@ bool furi_hal_crypto_decrypt(const uint8_t* input, uint8_t* output, size_t size)
+
+    return true;
+
+out_enc_err:
```
Nope

skotopes commented on firmware/targets/f7/furi_hal/furi_hal_crypto.c:
```diff
@@ -340,3 +382,377 @@ bool furi_hal_crypto_decrypt(const uint8_t* input, uint8_t* output, size_t size)
+
+    return true;
+
+out_enc_err:
```
Nope

skotopes commented on firmware/targets/f7/furi_hal/furi_hal_crypto.c:
```diff
@@ -340,3 +382,377 @@ bool furi_hal_crypto_decrypt(const uint8_t* input, uint8_t* output, size_t size)
+
+    state = true;
+
+out_enc_err:
```
Nope

skotopes commented on firmware/targets/f7/furi_hal/furi_hal_crypto.c:
```diff
@@ -24,6 +24,19 @@
 #define CRYPTO_AES_CBC (AES_CR_CHMOD_0)
 
+#define CRYPTO_AES_CTR (AES_CR_CHMOD_1)
+#define CRYPTO_CTR_IV_LEN 12
```
wrap in () and specify type

skotopes commented on firmware/targets/f7/furi_hal/furi_hal_crypto.c:
```diff
@@ -24,6 +24,19 @@
+#define CRYPTO_GCM_IV_LEN 12
+#define CRYPTO_GCM_CTR_LEN 4
+#define CRYPTO_GCM_TAG_LEN 16
+#define CRYPTO_GCM_PH_INIT 0U
```
wrap

skotopes commented on firmware/targets/f7/furi_hal/furi_hal_crypto.c:
```diff
@@ -276,6 +289,35 @@ bool furi_hal_crypto_store_unload_key(uint8_t slot) {
+}
+
+bool furi_hal_crypto_unload_key(void) {
+    if(!furi_hal_bt_is_alive()) {
```
same

skotopes commented on firmware/targets/f7/furi_hal/furi_hal_crypto.c:
```diff
@@ -340,3 +382,377 @@ bool furi_hal_crypto_decrypt(const uint8_t* input, uint8_t* output, size_t size)
+static bool wait_for_crypto(void) {
+    uint32_t countdown = CRYPTO_TIMEOUT;
+    while(!READ_BIT(AES1->SR, AES_SR_CCF)) {
+        if(LL_SYSTICK_IsActiveCounterFlag()) {
```
use furi cortex timer

twisted-pear replied:
How do I do that? I copied this logic from crypto_process_block(). Can you point me to an example, where it is done correctly?

skotopes replied:
Pretty much any place where you find `furi_hal_cortex_timer_get()`. It always ticking while mcu core clocked. Which is not always true for systick timer: there are some moment at early initialization stage where systick is not yet started.

skotopes commented:
And unit tests please ;-)

skotopes commented:
Undraft when ready. 

PS: I've updated documentation and renamed couple of our methods and was ready to merge it, before I saw goto and realized that it's too complex to be merged without unit tests. Really sorry for littering your branch with it.

Commit:
Clean up code for key (un)loading, GCM and CTR

- get rid of goto
- do not use furi_hal_bt_is_alive() when not using secure enclave
- give defines a type and wrap in ()

Commit:
Add unit test for CTR and GCM crypto

twisted-pear commented:
Added unit tests. I had some trouble getting the unit test version of the firmware running on my device, but the tests passed. I had to compile with COMPACT=1 DEBUG=0 or else fbt gave me a warning that the firmware wouldn't fit.

skotopes committed:
Merge branch 'dev' into crypto_enhancements

skotopes commented:
@twisted-pear normally we use SD-Card for test data. But I think we can merge it as is and then move data to sd later.

skotopes commented:
Also @hedger is currently working on API Version injection into defines, I think we can replace FURI_HAL_CRYPTO_ADVANCED_AVAIL with it.

Meanwhile I'll postpone this PR to next release(0.89).

skotopes commented:
And that's really great work that you've done here.

skotopes commented:
@twisted-pear so how about moving test data to sd?

twisted-pear commented:
> @twisted-pear so how about moving test data to sd?

I'll look into it, in the coming week. I have to look at other tests that do that.

Which file format can I use to store the test vectors?

skotopes committed:
Merge remote-tracking branch 'origin/dev' into crypto_enhancements

skotopes committed:
FuriHal: const in crypto unit tests, cortex timer for crypto operations timeouts

skotopes committed:
FuriHal: update crypto docs

skotopes commented:
Any reason why do we need `FURI_HAL_CRYPTO_ADVANCED_AVAIL`?

twisted-pear commented:
> Any reason why do we need `FURI_HAL_CRYPTO_ADVANCED_AVAIL`?

I have been using it in an app I'm developing to differentiate between FW versions that support GCM (where I use furi_hal_crypto*) and those that don't (where I use a GCM implementation that I copied from somewhere) at build time. If it is possible to get and compare the API version in the C preprocessor `FURI_HAL_CRYPTO_ADVANCED_AVAIL` is not really necessary. Otherwise, it is useful to maintain compatibility across different FW versions.

skotopes commented:
I'll merge this as is, but we are going to drop `FURI_HAL_CRYPTO_ADVANCED_AVAIL` closer to v1.0 replacing it with api version defines
